File: apo-core/trunk/src/main/java/com/uqbar/aop/javassit/builder/CtMethodBuilder.java
package com.uqbar.aop.javassit.builder;

import com.uqbar.aop.javassit.parser.JavassitParser;

import javassist.CtClass;
import javassist.CtMethod;
import javassist.CtNewMethod;

public class CtMethodBuilder {

	private String methodName;
	private int modifier;
	private CtClass returnType;
	private CtClass owner;
	private String body;
	private CtClass[] parametersType;

	public CtMethodBuilder withName(String name) {
		this.methodName = name;
		return this;
	}

	public CtMethodBuilder withModifier(int modifier) {
		this.modifier = modifier;
		return this;
	}

	public CtMethodBuilder withReturnType(CtClass returnType) {
		this.returnType = returnType;
		return this;
	}

	public CtMethodBuilder withOwner(CtClass owner) {
		this.owner = owner;
		return this;
	}

	public CtMethodBuilder witBody(String body) {
		this.body = body;
		return this;
	}

	public CtMethodBuilder withparameterClass(CtClass... parametersType) {
		this.parametersType = parametersType;
		return this;
	}
	
	public CtMethod build(){
		return createMethod(methodName, modifier, returnType, owner, body, parametersType);
	}
	
	private CtMethod createMethod(String methodName, int modifier, CtClass returnType, CtClass owner,
			String body, CtClass... parametersType)  {

		try {
			CtMethod method = CtNewMethod.make(
			         modifier,
			         returnType, 
			         methodName,//name
			         parametersType, //parameters type
			         new CtClass[]{},//exceptions
			         JavassitParser.parser(body),
			         owner);
			return method;
		} catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

}

File: apo-core/trunk/src/main/scala/com/uqbar/aop/javassit/builder/CtMethodBuilder.scala
package com.uqbar.aop.javassit.builder
import javassist.CtClass

import scala.collection.JavaConversions._
import javassist.CtMethod
import javassist.CtNewMethod
import com.uqbar.aop.javassit.parser.JavassistParser

class CtMethodBuilder {

  var methodName: String = _
  var modifier: Int = _
  var returnType: CtClass = _
  var owner: CtClass = _
  var parametersType: Array[CtClass] = Array()
  var body: String = _

  def withName(name: String)  = {this.methodName = name;this }

  def withModifier(modifier: Int) { this.modifier = modifier; this}

  def withReturnType(returnType: CtClass) = {this.returnType = returnType; this}

  def withOwner(owner: CtClass)= {this.owner = owner; this}

  def witBody(body: String)= {this.body = body; this}

  def withparameterClass(parametersType: CtClass*) = {this.parametersType = parametersType.toArray; this }

  def build() = createMethod(methodName, modifier, returnType, owner, body, parametersType)

  protected def createMethod(methodName: String, modifier: Int, returnType: CtClass, owner: CtClass,
    body: String, parametersType: Array[CtClass]): CtMethod = {

    return CtNewMethod.make(
      modifier,
      returnType,
      methodName, //name
      parametersType.toArray, //parameters type
      Array[CtClass](), //exceptions
      JavassistParser.parser(body),
      owner);
  }

}